front/lib/api/chat.ts

```diff
@@ -104,116 +111,102 @@ export async function getChatSession(
   };
 }
 
+export async function getChatMessage(
+  sId: string
+): Promise<ChatMessageType | null> {
+  const chatMessage = await ChatMessage.findOne({
+    where: { sId },
+  });
+  if (!chatMessage) {
+    return null;
+  }
+  const retrievedDocuments = await ChatRetrievedDocument.findAll({
+    where: {
+      chatMessageId: chatMessage.id,
+    },
+    order: [["score", "DESC"]],
+  });
+
+  return {
+    role: chatMessage.role,
+    message: chatMessage.message,
+    feedback: chatMessage.feedback,
+    sId: chatMessage.sId,
+    retrievals: retrievedDocuments.map((r) => {
+      return {
+        dataSourceId: r.dataSourceId,
+        sourceUrl: r.sourceUrl,
+        documentId: r.documentId,
+        timestamp: r.timestamp,
+        tags: r.tags,
+        score: r.score,
+        chunks: [],
+      };
+    }),
+  };
+}
+
 export function newChatSessionId() {
   const uId = new_id();
   return uId.slice(0, 10);
 }
 
-export async function storeChatSession(
-  sId: string,
-  owner: WorkspaceType,
-  user: UserType,
-  title: string | null,
-  messages: ChatMessageType[]
-): Promise<ChatSessionType> {
+export async function upsertChatMessage(
+  sessionId: number,
+  m: ChatMessageType,
+  sId: string
+): Promise<ChatMessageType> {
   return await front_sequelize.transaction(async (t) => {
-    // Start by cleaning up the previous chat session
-    const oldSession = await ChatSession.findOne({
-      where: {
-        workspaceId: owner.id,
-        sId,
-      },
-      transaction: t,
-    });
-
-    if (oldSession) {
-      // destroy all chat messages and retrived documents
-      await Promise.all(
-        (
-          await ChatMessage.findAll({
-            where: {
-              chatSessionId: oldSession.id,
-            },
-            transaction: t,
-          })
-        ).map((m) => {
-          return (async () => {
-            await ChatRetrievedDocument.destroy({
-              where: {
-                chatMessageId: m.id,
-              },
-              transaction: t,
-            });
-            await ChatMessage.destroy({
-              where: {
-                id: m.id,
-              },
-              transaction: t,
-            });
-          })();
-        })
-      );
-
-      // destroy the chat session
-      await ChatSession.destroy({
-        where: {
-          id: oldSession.id,
-        },
-        transaction: t,
-      });
-    }
-
-    const chatSession = await ChatSession.create(
+    const [message] = await ChatMessage.upsert(
```

Comment:
What is the exact semantic of upsert? How does sequelize know we care about `sId`?

Reply:
From the site, there is a specific postrges behaviour that would have probably fitted
```Note for Postgres users: If upsert payload contains PK field, then PK will be used as the conflict target. Otherwise first unique constraint will be selected as the conflict key.```
However, I think this is shaky and changed the logic following your comment

Reply:
Yes too much magic :)